
import libraries and data

In [1]:
import pandas as pd
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

In [2]:
df = pd.read_csv('/datasets/users_behavior.csv')

In [3]:
print(df.info())
print(df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None
   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0


    

Split data into training set, validation set and final test set with a 60/20/20 ratio 



In [4]:
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']

In [5]:
features_train, features_verify, target_train, target_verify = train_test_split(features, target, test_size=0.4, random_state=12345)


In [6]:
features_valid, features_test, target_valid, target_test = train_test_split(features_verify, target_verify, test_size=0.5, random_state=12345)


Good splitting of the data into `train`, `validation` and `test` datasets.


In [7]:


best_accuracy = 0
best_depth = None

for depth in range(1,11):
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model.fit(features_train, target_train)
    predictions = model.predict(features_verify)
    accuracy = accuracy_score(target_verify, predictions)

    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_depth = depth

print('Best Accuracy :', best_accuracy)
print('Best Depth :', best_depth)


    

Best Accuracy : 0.7908242612752722
Best Depth : 7




   Testing the DecisionTreeClassifier model for best accuracy and depth to compare with others. Best Accuracy was 0.79, which is higher than the minimum requirement. 



In [8]:

best_depth = 0 
best_score = 0
best_est = 0
for est in range(10,51,10):
    for depth in range(1,11):
        model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth)
        model.fit(features_train, target_train)
        score_valid = model.score(features_verify, target_verify)
        predictions = model.predict(features_verify)
      
        if score_valid > best_score:
            best_score = score_valid
            best_est = est
            best_depth = depth

print('best depth:', best_depth, 'best score:', best_score, 'best est:', best_est)   

best depth: 7 best score: 0.8032659409020217 best est: 30



Testing the RandomForestClassifier model to find best accuracy result. with a 0.803 it is higher than both the minimum requrement as well as the DecisionTreeClassifier Model.


In [9]:

model = LogisticRegression(random_state=12345)
model.fit(features_train, target_train)
train_score = model.score(features_train, target_train)
valid_score = model.score(features_verify, target_verify)
print('accuracy of training model:', train_score * 100)
print('accuracy of validation model:', valid_score * 100)


accuracy of training model: 71.31742738589212
accuracy of validation model: 69.751166407465




   LogisticRegression Model did not meet minimum requrements.

In [10]:


final_model = RandomForestClassifier(random_state=12345, n_estimators=30, max_depth=7)
final_model.fit(features_train, target_train)

final_score = final_model.score(features_test, target_test)

print('final training score:', final_score)



final training score: 0.8040435458786936


    

 Final Model with the higest accuracy was the RandomForestClassifier. Tested the final data against the training data with a final score of 0.804



In [11]:
smart_users = (df['is_ultra'] == 0).sum()
print(f"Smart plan users: {smart_users}")

ultra_users = (df['is_ultra'] == 1).sum()
print(f"Ultra plan users: {ultra_users}")

majority_class_accuracy = df['is_ultra'].value_counts(normalize=True).max()
print(majority_class_accuracy)


Smart plan users: 2229
Ultra plan users: 985
0.693528313627878


    
Final Sanity check of the model. Best Model was the RandomForestClassifier with an 80.4% accuracy versus the baseline of 69.35%


